### ISSUE TITLE
Drills cause massive sim speed slowdown

### ISSUE THREAD

**Issue by Aleks976** [KeenSoftwareHouse/SpaceEngineers#362]
Running 3770k CPU 4.5Ghz and 8GB 1600Mhz Ram, AMD R9 290 GPU.

Tested on both the current steam version (1.91) and the current source code compiled into 64bit mode. Using one drill or more causes the sim speed to go down from 1.00 to 0.50. CPU usage climbs very quickly too. (No drills = 13%, one drill = 44%, two drills = 70%, any more = 99%).

CPU usage climbs  ->  Sim speed drops  ->  FPS drops

**Comment by mccorkle** [comment]
I'm trying to write up a list of use cases that generate lag in SE. Can you help me with some more details on your world setup? 

For your world how many floating objects do you have allowed?  Does it just kill simspeed when you engage the drills in empty space, or when actually drilling a rock?  What is your view distance world setting? Which render engine (dx9 or dx11)? Are you hosting a multiplayer, of just in single player / offline?  

Could you also post a screenshot with the debug data (shift-f11 and ctrl-h)? I'd love to match it up to what I'm seeing.

**Comment by cubrr** [comment]
I did some CPU usage profiling in Visual Studio, here's your main CPU hog (click to open full size image):

[![image](https://cloud.githubusercontent.com/assets/6078228/8779634/f6415b40-2f0c-11e5-902c-7ff43ff2f2bf.png)](https://cloud.githubusercontent.com/assets/6078228/8779634/f6415b40-2f0c-11e5-902c-7ff43ff2f2bf.png)

**Comment by IhanaMies** [comment]
It is caused by this code snippet in MyDrillBase since it gets called every frame by every drill. Disabling it allows very large mining ships to drill without causing any major fps or simulation speed drops. Though disabling it stops drilling "smoke effect".

```
foreach (var entry in m_drilledMaterialBuffer)
            {
                somethingDrilled = (!collectOre || TryHarvestOreMaterial(entry.Key, hitPosition, entry.Value, onlyCheck)) || somethingDrilled;

                if (somethingDrilled && !onlyCheck)
                {
                    MyDebris.Static.CreateDirectedDebris(hitPosition,
                            MyUtils.GetRandomVector3Normalized(), 0.1f, 1, 0, MathHelper.Pi, 5, 1, 0.15f, entry.Key);
                }
            }
```

Getting it called by UpdateAfterSimulation10() would probably fix most of the lag and simulation speed drops and still produces the smoke effect.

**Comment by joemorin73** [comment]
The "smoke effect" being the debris from what I see.  Anyone trying this idea?

**Comment by IhanaMies** [comment]
I did get some results with a simple +=1 counter when MyDebris.Static.CreateDirectedDebris and CreateParticles in line 343 gets called. I'm experimenting with different counter resetting numbers to see if I could find a good normal for them.

MyDebris.Static.CreateDirectedDebris works like this:
```
        public void CreateDirectedDebris(Vector3 sourceWorldPosition,
                                         Vector3 offsetDirection,
                                         float minSourceDistance,
                                         float maxSourceDistance,
                                         float minDeviationAngle,
                                         float maxDeviationAngle,
                                         int debrisPieces,
                                         float initialSpeed,
                                         float scale,
                                         MyVoxelMaterialDefinition material)
        {
            ProfilerShort.Begin("Create directed debris");
            MyDebug.AssertDebug(debrisPieces > 0);
            for (int i = 0; i < debrisPieces; ++i)
            {
                var newObj = CreateVoxelDebris();
                if (newObj == null)
                {
                    break; // no point in continuing
                }

                float dist = MyUtils.GetRandomFloat(minSourceDistance, maxSourceDistance);
                float angleX = MyUtils.GetRandomFloat(minDeviationAngle, maxDeviationAngle);
                float angleY = MyUtils.GetRandomFloat(minDeviationAngle, maxDeviationAngle);
                var rotation = Matrix.CreateRotationX(angleX) * Matrix.CreateRotationY(angleY);
                var deviatedDir = Vector3.Transform(offsetDirection, rotation);
                var startPos = sourceWorldPosition + deviatedDir * dist;
                var initialVelocity = deviatedDir * initialSpeed;
                (newObj.Debris as MyDebrisVoxel.MyDebrisVoxelLogic).Start(startPos, initialVelocity, scale, material);
            }
            ProfilerShort.End();
        }
```
Volume or mass are not defined here. Does this mean these are only visual rock pieces flying off and getting deleted after some time?

**Comment by IhanaMies** [comment]
With the changes I've made I could drill constantly with a 169 drill ship at 1.0 m/s fps being around 20, frametime 50-100ms and simulation speed around 50 to 60.

Normally my fps was around 5, frametime 100-400 and simulation speed 30-45.

Have to experiment more.

**Comment by joemorin73** [comment]
It definitely sounds like an improvement.

**Comment by mexmer** [comment]
can you check  how much it affects perfomance it you move it to updateaftersimulation10 (which shlould be every 10th frame)

**Comment by Aleks976** [comment]
First, sorry about 21:9 instead of 16:9 screenshots :)

I am using a 20km view distance. Single player offline. I have a 30 object limit on this world. I tested in both DX9 and DX11

DX11 Drills Off
![2015-07-20_00001](https://cloud.githubusercontent.com/assets/13395306/8785783/1190e39e-2ef8-11e5-956f-a6e51a0df431.jpg)

DX11 Drills On
![2015-07-20_00002](https://cloud.githubusercontent.com/assets/13395306/8785793/2496d8cc-2ef8-11e5-9baf-59d1b0f91480.jpg)

DX11 Drilling
![2015-07-20_00004](https://cloud.githubusercontent.com/assets/13395306/8785803/2c0e4400-2ef8-11e5-9fc8-ac14f999bed7.jpg)

DX9 Drills On
![2015-07-20_00005](https://cloud.githubusercontent.com/assets/13395306/8785812/3d271168-2ef8-11e5-8766-a8392ed1b9c5.jpg)

DX9 Drilling
![2015-07-20_00006](https://cloud.githubusercontent.com/assets/13395306/8785817/42ab23f4-2ef8-11e5-8448-3757f055f53b.jpg)

**Comment by joemorin73** [comment]
Is that the baseline or with the UpdateSimulation10?

**Comment by mccorkle** [comment]
@joemorin73 for Aleks976's screen shots, he said it was baseline. 

@IhanaMies is it easy for us to get your build of the SE binary to test the changes on some different machines?

**Comment by devu** [comment]
Interesting find, but I will add into this. On our server with limit of
flying objects set to 64 we were able to reproduce consistently the sim
speed drop.
However what was consistent, it wasn't a process of mining or generating
derbies during mining at all, but when miner was leaving drilling area.

How that would apply?
From the profiling screen grab send by Joona looks like ReadRange sims to
be pretty intensive task in the process.

**Comment by IhanaMies** [comment]
@mccorkle here: https://www.dropbox.com/s/c9aqrsidcak52a4/MyDrillBase.cs?dl=0

**Comment by mexmer** [comment]
Ok, i checked code, and seems that function, that call this function is called from UpdateAfterSimulation10(), also from Shoot().
If you have drill powered, it can create debris only once per 10 frames ... disregard my request then.

**Comment by Aleks976** [comment]
I would like to update this thread: this is not a mining issue. It is a issue with voxels being destroyed. I just tested by firing missiles at asteroids: as more missiles are fired, the cpu usage gets high and higher. At 15+ rocket pods, cpu usage hits 100% and sim speed starts dropping.

**Comment by mexmer** [comment]
that would make sense, if debris is created upon voxel destruction, and if debris is culprit.

**Comment by cubrr** [comment]
> From the profiling screen grab send by Joona looks like ReadRange sims to
> be pretty intensive task in the process.

Definitely. The [`OctreeStorage.ReadRange`](https://github.com/KeenSoftwareHouse/SpaceEngineers/blob/b969b03f11be41448f7dd85c0bdbd45536d1e420/Sources/Sandbox.Game/Engine/Voxels/MyOctreeStorage.cs#L712) method looks pretty heavy.

**Comment by Tyrsis** [comment]
If this is the server, debris should not even spawned.  They are mostly effects that don't really need to be tracked.  Spawn them on the client to keep the effect, but drop them on the server.

**Comment by joemorin73** [comment]
@Tyrsis Doesn't this need to be done carefully?  The Server not also mean the hosting machine in a non-DS?

Is there a flag that says DS?

**Comment by Tyrsis** [comment]
You'd be surprised what you can get away with on the server.  This really has no place being on the server.  Just like the particle manager, this is mostly just an effect that doesn't need to even be run. 

You could probably just do a Sync.IsServer check, and if it's true, don't create debris.  It'll probably still create the debris on the client that does no damage, but who cares.  The debris is not the floating objects, it's just the effect of those little stones flying out then disappearing.

**Comment by Tyrsis** [comment]
After quickly looking, I don't see anything syncing from server -> client in regards to these entities, so I highly doubt removing it from the server would even cause you to lose the visual, as the visual isn't linked to anything on the server sync wise.  The only thing you'd lose is the bit of physics these things have, but I have never seen them do damage to anything, so I don't think it even matters.

**Comment by joemorin73** [comment]
If it's going through the sync messaging, I can agree.  To be safe, it might be best to use the IsDedicated in regards to the rendering on DS.

But this isn't what this issue is about since this discussion can tangent quickly.

**Comment by Tyrsis** [comment]
Well the issue is that debris causes lag on the client, but it also creates just as much lag on the server, which can be an issue.  And while a solution here might optimize one mining ship, it may not fix when you have 10 players mining at once.  It's better to just drop this mechanic from the dedicated server completely.  I don't think it's too much of a tangent.

**Comment by devu** [comment]
I would agree on that.

Still the heaviest thing there to investigate is OctreeStorage.ReadRange
<https://github.com/KeenSoftwareHouse/SpaceEngineers/blob/b969b03f11be41448f7dd85c0bdbd45536d1e420/Sources/Sandbox.Game/Engine/Voxels/MyOctreeStorage.cs#L712>
Notice there is a lag and sim speed drop not only during a drilling
operation but when you leaving mining area.

**Comment by joemorin73** [comment]
@tyrsis, agreed.

**Comment by Tyrsis** [comment]
ReadRange is how the voxel is read.  I'm not entirely certain what you're talking about in terms of "leaving the area", but I believe what you're experience is not related to this, and is actually a much different issue.

Where are you seeing ReadRange called?

**Comment by cubrr** [comment]
Tyrsia, we're referring to this Visual Studio performance profiling report:
![](https://cloud.githubusercontent.com/assets/6078228/8779634/f6415b40-2f0c-11e5-902c-7ff43ff2f2bf.png)
I'm not sure what the _External Code_ section is about.
On Jul 22, 2015 00:12, Tyrsis <notifications@github.com> wrote:ReadRange is how the voxel is read.  I'm not entirely certain what you're talking about in terms of "leaving the area", but I believe what you're experience is not related to this, and is actually a much different issue.

Where are you seeing ReadRange called?

**Comment by devu** [comment]
To clarify, the "leaving area" Is not exactly what I said. I said "when you
(are) leaving mining area" / mined asteroid. Because we talking about
massive sim speed slowdown due to mining operation here.
but I believe it has something to do to mining but the effect is delayed
somehow, I am able to reproduce that each time. I believe those 2 things
are related.

If that's another issue well... could you point me to the direction?

**Comment by Tyrsis** [comment]
Yes look at that profile closely.  See the "ParallelTasks.WorkItem" ?  That means it's being done in a thread.  Also that is for the ore detector.  

That is going to cause CPU usage on another core other than the game thread.  It's nothing to worry about (unless you have 1 CPU??).  This is actually one of very few parts of the game properly threaded.

**Comment by cubrr** [comment]
I see, that's very interesting. Thank you for clearing it up.

**Comment by alex-davidson** [comment]
The slowdown when leaving the asteroid's vicinity appears to be caused by meshes being requested on the main thread, possibly because the colliding objects don't get added to the 'nearby' list and therefore never cause precalc work to be queued on other threads. As soon as the number of grids near the asteroid falls to zero, all those meshes get handled on the main thread.

Modifying MyVoxelPhysicsBody to include MyFloatingObjects in m_nearbyEntities (handlers for CollidableRemoved and CollidableAdded) does seem to solve the problem at the expense of increased overall CPU usage. A more efficient solution is doubtless possible but currently beyond my knowledge.

Currently testing a possible workaround.

**Comment by alex-davidson** [comment]
Workaround: place a single small-ship landing gear near the asteroid (<500m?). This forces a precalc job.

**Comment by paradineai** [comment]
I have an idea for the devs to try. What about creating a zero collision dummy (like a ship block that you can not collide with) so that when you mine an asteroid once it has been deemed altered (the same way they do for the procedural asteroids to tell the game if it should be saved or not) then it adds the dummy to the game near the asteroid (<500m) so it forces the game to do precalc for the asteroid but the dummy does not show up in the save files or at the least takes up as little data as possible.

**Comment by alex-davidson** [comment]
That would add complex behaviour to the system to work around a bug, rather than just fixing the buggy behaviour (ie. it's a hack :P). A simpler way of doing the same thing would be just to queue a precalc job as long as floating objects are nearby, instead of adding a dummy object to try to get the rest of the code to do this.

But as it turned out, the lack of precalc doesn't have anything to do with the bug. The problem was that when there are no nearby objects the asteroid's shape information is discarded to save memory (the lack of precalc is just another side-effect). This shape is then recalculated every time the physics engine detects a collision ie. hundreds of times a second.

The fix is therefore to keep the shape in memory while any collidable object exists nearby. This does mean that this memory cannot be freed as aggressively but it does avoid needlessly recalculating that information every sim update. A possible improvement might be to reduce the shape's detail level in these circumstances, so it takes up less memory.

Once the objects all despawn, the shape will be removed as usual and the memory will be freed.
